### PR TITLE
CART-89 build: Add gofmt to code checking

### PR DESCRIPTION
**Pull request by ashleypittman**
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

File: src/control/cmd/daos/pool.go [hunk 1]
```diff
@@ -79,7 +79,7 @@ func (cmd *poolBaseCmd) connectPool(flags C.uint) error {
 		cmd.log.Debugf("connecting to pool: %s", cmd.PoolID().Label)
 		rc = C.daos_pool_connect2(cLabel, cSysName, flags,
 			&cmd.cPoolHandle, &poolInfo, nil)
-		if rc == 0 {
+		if (rc == 0) {
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/control/cmd/daos/pool.go]: This is an example of the comment that Jenkins will make on PRs if there are problems.

**Review by tanabarr — DISMISSED** [review]

**Review by mjmac — COMMENTED** [review]
Hmm. I assume the goal here is to avoid penalizing developers who haven't configured their editor to run gofmt on save, or otherwise run it before committing. I can understand the desire, but I'm not sure I agree with the end result of adding this band-aid, as the committed code could be formatted incorrectly.

This means that developers could get unformatted code when they pull the repo, which means that their local testing will fail unless they fix the formatting locally, and then they have to include unrelated formatting changes when they create a commit (or else revert the formatting changes).

I think a better solution to this problem would involve a local git hook and/or better documentation for developers so that they understand what they need to do in order to avoid having their PR fail in unit test. Additionally, I have long wanted to pull the Go unit tests out into their own stage, but just haven't had the time to figure out how to make that work. If nothing else, it would make Go fmt/test failures much easier to spot and fix.

**Comment by ashleypittman** [comment]
> Hmm. I assume the goal here is to avoid penalizing developers who haven't configured their editor to run gofmt on save, or otherwise run it before committing. I can understand the desire, but I'm not sure I agree with the end result of adding this band-aid, as the committed code could be formatted incorrectly.
> 
> This means that developers could get unformatted code when they pull the repo

The Unit Tests will still fail on badly formatted code, the intention here is absolutely not to reduce code quality in committed code.

> Additionally, I have long wanted to pull the Go unit tests out into their own stage, but just haven't had the time to figure out how to make that work. If nothing else, it would make Go fmt/test failures much easier to spot and fix.

This change is really a step in this direction, the change will be so that if people who don't have the save/commit hooks in place will be notified of the failure earlier, and in a format that's easier to understand.

**Comment by mjmac** [comment]
> 
> This change is really a step in this direction, the change will be so that if people who don't have the save/commit hooks in place will be notified of the failure earlier, and in a format that's easier to understand.

I am all for that, but I don't see how this PR achieves that goal. The gofmt invocation in `ci/patch_src_in_place` just does an automatic reformat -- it doesn't fail if the code needs to be formatted. Do you need to do something more like this?
```sh
find src/control -name '*.go' -and -not -path '*vendor*' -print0 | xargs -0 gofmt -d
```

**Comment by ashleypittman** [comment]
> > This change is really a step in this direction, the change will be so that if people who don't have the save/commit hooks in place will be notified of the failure earlier, and in a format that's easier to understand.
> 
> I am all for that, but I don't see how this PR achieves that goal. The gofmt invocation in `ci/patch_src_in_place` just does an automatic reformat -- it doesn't fail if the code needs to be formatted. Do you need to do something more like this?
> 
> ```shell
> find src/control -name '*.go' -and -not -path '*vendor*' -print0 | xargs -0 gofmt -d
> ```

It's weird but logical.  The patch_src_in_place script is intended as a way of adding support for tools which can reformat src, exactly like gofmt or checkpatch.  What happens is that the script is run, applies any corrections that it can and then there's code in pipeline-lib that translates these into comments into the PR, that github presents as suggestions.  Some like codespell and go we expect the user to always pick up and silence but some of the checkpatch ones are advisory so this step doesn't fail if there are suggestions however.

**Review by mjmac — DISMISSED** [review]
Thanks for the explanation, I get it now. LGTM

**Review by mjmac — APPROVED** [review]

**Review by tanabarr — APPROVED** [review]

**Pull request merged**